{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 8. Finalize Table S8 by adding sequences"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Goals:\n",
    "1. Get start/end coords of rhodopsins -> add to Table S8\n",
    "2. Add faa and CDs for rhodopsins and flanking flotillins/ferredoxins->  Table S8\n",
    "3. *Almost forgot!* Add 'retinal_binding_residue' and 'has_retinal_binding' -> Table S8 "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Starting\n",
      "### I.  Subsetting dataset to work only with rhodopsin-bearing SAGs.\n",
      "\tOtherwise, using full GORG Dark takes 10 min for each combined dataframe to load.\n",
      "\tDropping all genes not on the 1079 contigs with rhodopsin\n",
      "\t39133 annotation rows remaining.\n",
      "\n",
      "### II.  Reading tables -> dataframes. Fastas -> dictionaries.\n",
      "\t1093 rhodopsins in GORG Dark\n",
      "\t422 rhods flanking ferredoxin\n",
      "\t391 rhods flanking flotillin\n",
      "\tLoaded 38369 prots from rhodopsin-bearing contigs\n",
      "\t757689 CDS loaded from DRAM\n",
      "\t757689 FAA loaded from DRAM\n",
      "\n",
      "### III.    Finding the flotillin and ferredoxin seqs that neighbored rhodopsin\n",
      "\tFound 422 proteins annotated as ferredoxin, that are next to rhodopsin\n",
      "\tFound 372 proteins annotated as flotillin, that are next to rhodopsin\n",
      "\n",
      "### VI.   Adding rhodopsin, flotillin, and ferredoxin, sequences to table...\n",
      "\n",
      "\tRhodopsin\n",
      "\t1093 proteins to find\n",
      "\t1093 found\n",
      "\t1093 rows of the rhodopsin metadata table now have a protein added for rhodopsin\n",
      "\t1093 CDSs to find\n",
      "\t1093 found\n",
      "\t1093 rows of the rhodopsin metadata table now have a CDS added for rhodopsin\n",
      "\n",
      "\tFerredoxin\n",
      "\t422 proteins to find\n",
      "\t422 found\n",
      "\t422 rows of the rhodopsin metadata table now have a protein added for ferredoxin\n",
      "\t422 CDSs to find\n",
      "\t422 found\n",
      "\t422 rows of the rhodopsin metadata table now have a CDS added for ferredoxin\n",
      "\n",
      "\tFlotillin\n",
      "\t372 proteins to find\n",
      "\t372 found\n",
      "\t372 rows of the rhodopsin metadata table now have a protein added for flotillin\n",
      "\t372 CDSs to find\n",
      "\t372 found\n",
      "\t372 rows of the rhodopsin metadata table now have a CDS added for flotillin\n",
      "\n",
      "#### V.    Loading start/end coordinates for rhodopsin genes\n",
      "\n",
      "\n",
      "### VI.    Add operon info to table\n",
      "\n",
      "### VII.   Polishing and saving\n",
      "\tSaving...\n",
      "\tDone.\n",
      "\n",
      "### VIII.   Adding in the retinal_binding info that Tianyi forgot!\n",
      "adding 2 columns\n",
      "Writing out updated table\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from glob import glob\n",
    "import os\n",
    "from os.path import exists\n",
    "import shutil\n",
    "from Bio import SeqIO\n",
    "import sys\n",
    "\n",
    "# Output\n",
    "PATH_out = \"../Table_S8_Rhodopsin_sequences_metadata_and_gene_neighborhoods.tsv\"\n",
    "\n",
    "# Inputs\n",
    "PATH_rhod_meta = \"../7_integrate_alphafold_assessment/GORG_dark_rhodopsin_metadata_with_phylogeny_and_alphafold3.tsv\" # This is the table that we're improving here\n",
    "# Retinal binding info that Tianyi forgot to add til this step\n",
    "PATH_retinal_CSV = \"./input_retinal_binding_residue/retinal_binding_residue.csv\"\n",
    "PATTERN_operons_CSV = \"../7_integrate_alphafold_assessment/rhodopsin_operon_presence_absence*csv\"  # Using a pattern because the date in the filename can change\n",
    "PATH_annot_rhodopsin_neighbors_TSV = \"../4_reannotate_dramAAs_based_on_foldseek_annot/reannot_all_prot_from_contigs_with_rhodopsins.tsv\"\n",
    "PATH_rhod_neighbors_FAA = \"../4_reannotate_dramAAs_based_on_foldseek_annot/reannot_Erank_proteins_from_rhodopsin_contigs.fasta\"\n",
    "PATH_dram_all_annots = \"../0_get_DRAM_AAs_for_SAGs_w_rhodopsins/dram_926_SAGs_combined.csv\"\n",
    "PATH_dram_all_CDS = \"../0_get_DRAM_AAs_for_SAGs_w_rhodopsins/dram_926_SAGs_combined.fna\"\n",
    "PATH_dram_all_FAA = \"../0_get_DRAM_AAs_for_SAGs_w_rhodopsins/dram_926_SAGs_combined.faa\"\n",
    "\n",
    "print(\"Starting\")\n",
    "### Add sequence columns to the rhodopsin metadata table\n",
    "# Results in col called either 'rhodopsin_protein(DRAM)', 'rhodopsin_CDS(DRAM)', 'flanking_ferredoxin_protein(DRAM)', 'flanking_ferredoxin_CDS(DRAM)', 'flanking_flotillin_protein(DRAM)', or 'flanking_flotillin_CDS(DRAM)' \n",
    "def add_col_with_sequence(DF_in,CDS_or_protein,rhodopsin_ferredoxin_flotillin):\n",
    "    \n",
    "    # Check input arguments\n",
    "    if CDS_or_protein not in ['CDS','protein']:\n",
    "        sys.exit(\"ERROR, argument 2 of add_col_with_sequence() must be 'CDS' or 'protein'\")\n",
    "    if rhodopsin_ferredoxin_flotillin not in ['rhodopsin', 'ferredoxin', 'flotillin']:\n",
    "        sys.exit(\"ERROR, argument 3 of add_col_with_sequence() must be 'rhodopsin', 'ferredoxin', 'flotillin'\")\n",
    "    \n",
    "    # What's the new column's name?\n",
    "    newColName = rhodopsin_ferredoxin_flotillin + \"_\" + CDS_or_protein + \"(DRAM)\"\n",
    "    if rhodopsin_ferredoxin_flotillin in ['ferredoxin', 'flotillin']:\n",
    "        newColName = \"flanking_\" + newColName\n",
    "        \n",
    "    # Choose the lists/dicts that will be searched against eachother e.g. ( rhodopsin vs. CDS ) | ( ferredoxin vs. protein ) | etc.\n",
    "    # Also, for flotillins and ferredoxins, need DICT that says which rhodopsinID each of them neighbors\n",
    "    \n",
    "    ## query: list of seqIDs\n",
    "    if rhodopsin_ferredoxin_flotillin == 'rhodopsin':\n",
    "        LIST_IDS_to_find = LIST_rhodopsin_IDs\n",
    "    if rhodopsin_ferredoxin_flotillin == 'ferredoxin':\n",
    "        LIST_IDS_to_find = LIST_ferredoxin_IDs\n",
    "    if rhodopsin_ferredoxin_flotillin == 'flotillin':\n",
    "        LIST_IDS_to_find = LIST_flotillin_IDs\n",
    "    \n",
    "    ## target: DICT of seqID2seqRecord (in Biopython format)\n",
    "    if CDS_or_protein == 'CDS':\n",
    "        DICT_to_search = DICT_dramCDS_ID2string\n",
    "    if CDS_or_protein == 'protein':\n",
    "        DICT_to_search = DICT_dramFAA_ID2string\n",
    "    \n",
    "    ## neighbor: DICT of rhodID2neighborID\n",
    "    if rhodopsin_ferredoxin_flotillin == 'ferredoxin':\n",
    "        DICT_rhod2neighbor = DICT_rhodID_2_ferredoxinID\n",
    "    if rhodopsin_ferredoxin_flotillin == 'flotillin':\n",
    "        DICT_rhod2neighbor = DICT_rhodID_2_flotillinID\n",
    "        \n",
    "    # Check that we'll be able to find what we need (i.e. that every query SeqID has a matching key in the target DICT)\n",
    "    print(\"\\t\"+str(len(LIST_IDS_to_find)) + \" \" + CDS_or_protein + \"s to find\")\n",
    "    count_found = 0\n",
    "    for ID in LIST_IDS_to_find:\n",
    "        if ID in list(DICT_to_search.keys()):\n",
    "            count_found += 1\n",
    "    print(\"\\t\"+str(count_found) + \" found\")\n",
    "\n",
    "    #### Finally, add the new column and populate it with the sequence:\n",
    "    _DF = DF_in\n",
    "    \n",
    "    # Every rhodopsin row will have a protein and CDS to retrieve\n",
    "    if rhodopsin_ferredoxin_flotillin == 'rhodopsin':\n",
    "        _DF[newColName] = DF_in['rhodopsin_gene'].map(DICT_to_search)\n",
    "    \n",
    "    \n",
    "    #But *NOT* every rhodopsin row will have an associated ferredoxin or flotillin\n",
    "    #So we need to use \"where\" logic,\n",
    "    if rhodopsin_ferredoxin_flotillin in ['ferredoxin', 'flotillin']:\n",
    "        \n",
    "        _DF[newColName] = np.where(DF_in['rhodopsin_gene'].isin(list(DICT_rhod2neighbor.keys())),\n",
    "                                   DF_in['rhodopsin_gene'].map(DICT_rhod2neighbor).map(DICT_to_search),\n",
    "                                   #DF_in['rhodopsin_gene'].apply(lambda x: \">\"+ str(DICT_to_search[DICT_rhod2neighbor[x]].id) + \"\\n\" + str(DICT_to_search[DICT_rhod2neighbor[x]].seq) ),\n",
    "                                   np.nan)\n",
    "        ### ^ What does this \"where\" logic mean?\n",
    "        #                 np.where( **IF** this row's rhodopsin has a neighbor \n",
    "        #                          **THEN**  populate this cell with that neighbor's seq like \">SeqID/nATATATAGGTATATATAT\"\n",
    "        #                          **ELSE**  leave cell empty ) \n",
    "                        \n",
    "            \n",
    "            \n",
    "    DICT_seq_populated = ~_DF[newColName].isna().value_counts() # Count how many arent empty\n",
    "    print(\"\\t\"+str(abs(DICT_seq_populated[False])-1) + \" rows of the rhodopsin metadata table now have a \"+CDS_or_protein+\" added for \"+rhodopsin_ferredoxin_flotillin)\n",
    "    \n",
    "    return _DF\n",
    "\n",
    "\n",
    "\n",
    "print(\"### I.  Subsetting dataset to work only with rhodopsin-bearing SAGs.\")\n",
    "print(\"\\tOtherwise, using full GORG Dark takes 10 min for each combined dataframe to load.\")\n",
    "\n",
    "DF_rhod_meta = pd.read_csv(PATH_rhod_meta, sep=\"\\t\")\n",
    "\n",
    "DF_annot = pd.read_csv(PATH_dram_all_annots)\n",
    "\n",
    "DF_annot = DF_annot.rename(columns={'...1':'rhodopsin_gene'})\n",
    "\n",
    "LIST_contigs_with_rhodopsin = DF_rhod_meta['rhodopsin_scaffold'].unique().tolist()\n",
    "print(\"\\tDropping all genes not on the \" +str(len(LIST_contigs_with_rhodopsin)) +\" contigs with rhodopsin\")\n",
    "DF_annot = DF_annot.loc[DF_annot['scaffold'].isin(LIST_contigs_with_rhodopsin)]\n",
    "print(\"\\t\"+str(len(DF_annot))+ \" annotation rows remaining.\\n\")\n",
    "\n",
    "DF_annot = DF_annot.rename(columns={'...1':'rhodopsin_gene'})\n",
    "\n",
    "\n",
    "\n",
    "print(\"### II.  Reading tables -> dataframes. Fastas -> dictionaries.\")\n",
    "    \n",
    "LIST_operon_CSV = glob(PATTERN_operons_CSV)\n",
    "if len(LIST_operon_CSV) != 1:\n",
    "    sys.exit(\"ERROR. Could not find a .csv file beginning with 'rhodopsin_operon_presence_absence' in working directory OR found too many of them\")\n",
    "else:\n",
    "    DF_operons = pd.read_csv(LIST_operon_CSV[0])\n",
    "    \n",
    "LIST_rhodopsin_geneIDs = DF_rhod_meta['rhodopsin_gene'].unique().tolist()\n",
    "LIST_rhods_flanking_flotillin = list(DF_operons.loc[DF_operons[\"3' flotillin\"]==1]['rhodopsin_gene'])\n",
    "LIST_rhods_flanking_ferredoxin = list(DF_operons.loc[DF_operons[\"3' 2Fe-2S binding dom.\"]==1]['rhodopsin_gene'])\n",
    "print(\"\\t\"+str(len(LIST_rhodopsin_geneIDs)) + \" rhodopsins in GORG Dark\")\n",
    "print(\"\\t\"+str(len(LIST_rhods_flanking_ferredoxin)) + \" rhods flanking ferredoxin\")\n",
    "print(\"\\t\"+str(len(LIST_rhods_flanking_flotillin)) + \" rhods flanking flotillin\")\n",
    "\n",
    "### A DRAM annotation table with columns 'locus tag', 'description', 'new_rankE_description'\n",
    "### LACKS gene start and stop positions\n",
    "DF_simple_annot = pd.read_csv(PATH_annot_rhodopsin_neighbors_TSV, sep=\"\\t\")\n",
    "\n",
    "DICT_proteins = SeqIO.index(PATH_rhod_neighbors_FAA, \"fasta\")\n",
    "LIST_protIDs_from_rhod_contigs = list(DICT_proteins.keys())\n",
    "print(\"\\tLoaded \"+str(len(LIST_protIDs_from_rhod_contigs))+ \" prots from rhodopsin-bearing contigs\")\n",
    "\n",
    "### Load protein and CDS fastas to search for sequences\n",
    "DICT_dramCDS_ID2record = SeqIO.index(PATH_dram_all_CDS, \"fasta\")\n",
    "DICT_dramCDS_ID2string = {}\n",
    "# Convert it from biopython Seq.record object to string\n",
    "for key in DICT_dramCDS_ID2record.keys():\n",
    "    record = DICT_dramCDS_ID2record[key]\n",
    "    STR_record = \">\" + record.id + \"\\n\" + record.seq\n",
    "    DICT_dramCDS_ID2string[key]=STR_record\n",
    "LIST_IDS_dramCDS = [ i for i in DICT_dramCDS_ID2record.keys() ]\n",
    "print(\"\\t\"+str(len(LIST_IDS_dramCDS))+ \" CDS loaded from DRAM\")\n",
    "\n",
    "DICT_dramFAA_ID2record = SeqIO.index(PATH_dram_all_FAA, \"fasta\")\n",
    "DICT_dramFAA_ID2string = {}\n",
    "\n",
    "# Convert it from biopython Seq.record object to string\n",
    "for key in DICT_dramFAA_ID2record.keys():\n",
    "    record = DICT_dramFAA_ID2record[key]\n",
    "    STR_record = \">\" + record.id + \"\\n\" + record.seq\n",
    "    DICT_dramFAA_ID2string[key]=STR_record\n",
    "LIST_IDS_dramFAA = [ i for i in DICT_dramFAA_ID2record.keys() ]\n",
    "print(\"\\t\"+str(len(LIST_IDS_dramFAA))+ \" FAA loaded from DRAM\")\n",
    "\n",
    "\n",
    "\n",
    "print(\"\\n### III.    Finding the flotillin and ferredoxin seqs that neighbored rhodopsin\")\n",
    "\n",
    "LIST_ferredoxin_records = []\n",
    "DICT_rhodID_2_ferredoxinID = {} # To track which rhodopsin neighbors which ferredoxin\n",
    "\n",
    "LIST_flotillin_records = []\n",
    "DICT_rhodID_2_flotillinID = {} # To track which rhodopsin neighbors which flotillin\n",
    "\n",
    "for rhodID in LIST_rhodopsin_geneIDs:\n",
    "    reformat_rhodID = rhodID[0:11] + rhodID\n",
    "    record_upstream = ''\n",
    "    record_downstream = ''\n",
    "    position_upstream = ''\n",
    "    position_downstream = ''\n",
    "    # e.g. 'AG-538-A02_contigs_AG-538-A02_NODE_4_1' -> 'AG-538-A02_A02_AG-538-A02_contigs_AG-538-A02_NODE_4_1'\n",
    "    if reformat_rhodID in LIST_protIDs_from_rhod_contigs:\n",
    "        position_upstream = LIST_protIDs_from_rhod_contigs.index(reformat_rhodID)-1 # e.g. 30\n",
    "        seqID_upstream = LIST_protIDs_from_rhod_contigs[position_upstream] # e.g. AG-538-B02_AG-538-B02_contigs_AG-538-B02_NODE_14_3\n",
    "        record_upstream = DICT_proteins[seqID_upstream]\n",
    "        if rhodID in LIST_rhods_flanking_ferredoxin:\n",
    "            if 'ferredoxin' in str(record_upstream.description) or 'PF00111' in str(record_upstream.description):\n",
    "                # Check that I'm only taking ferredoxins from SAGs with rhodopsins in the 2Fe-2S clade\n",
    "                #if seqID_upstream[0:10] in LIST_SAGS_of_interest:\n",
    "                LIST_ferredoxin_records.append(record_upstream)\n",
    "                DICT_rhodID_2_ferredoxinID[rhodID]=record_upstream.id[11:]\n",
    "        if rhodID in LIST_rhods_flanking_flotillin:\n",
    "            if 'lotillin' in str(record_upstream.description):\n",
    "                LIST_flotillin_records.append(record_upstream)\n",
    "                DICT_rhodID_2_flotillinID[rhodID]=record_upstream.id[11:]\n",
    "        \n",
    "        position_downstream = LIST_protIDs_from_rhod_contigs.index(reformat_rhodID)+1\n",
    "        seqID_downstream = LIST_protIDs_from_rhod_contigs[position_downstream]\n",
    "        record_downstream = DICT_proteins[seqID_downstream]\n",
    "        if rhodID in LIST_rhods_flanking_ferredoxin:\n",
    "            if 'ferredoxin' in str(record_downstream.description) or 'PF00111' in str(record_downstream.description):\n",
    "                # Check that I'm only taking ferredoxins from SAGs with rhodopsins in the 2Fe-2S clade\n",
    "                #if seqID_downstream[0:10] in LIST_SAGS_of_interest:\n",
    "                LIST_ferredoxin_records.append(record_downstream)\n",
    "                DICT_rhodID_2_ferredoxinID[rhodID]=record_downstream.id[11:]\n",
    "        if rhodID in LIST_rhods_flanking_flotillin:\n",
    "            if 'lotillin' in str(record_downstream.description):\n",
    "                LIST_flotillin_records.append(record_downstream)\n",
    "                DICT_rhodID_2_flotillinID[rhodID]=record_downstream.id[11:]\n",
    "  \n",
    "\n",
    "LIST_rhodopsin_IDs = LIST_rhodopsin_geneIDs\n",
    "LIST_ferredoxin_IDs = [record.id[11:] for record in LIST_ferredoxin_records] # e.g. 'AG-538-A02_AG-538-A02_contigs_AG-538-A02_NODE_4_9' -> 'AG-538-A02_contigs_AG-538-A02_NODE_4_9'\n",
    "LIST_flotillin_IDs = [record.id[11:] for record in LIST_flotillin_records]\n",
    "\n",
    "                \n",
    "print(\"\\tFound \"+str(len(LIST_ferredoxin_records)) + ' proteins annotated as ferredoxin, that are next to rhodopsin')\n",
    "print(\"\\tFound \"+str(len(LIST_flotillin_records)) + ' proteins annotated as flotillin, that are next to rhodopsin')\n",
    "\n",
    "\n",
    "\n",
    "print(\"\\n### VI.   Adding rhodopsin, flotillin, and ferredoxin, sequences to table...\")\n",
    "DF = DF_rhod_meta\n",
    "print(\"\\n\\tRhodopsin\")\n",
    "DF = add_col_with_sequence(DF,'protein','rhodopsin')\n",
    "DF = add_col_with_sequence(DF,'CDS','rhodopsin')\n",
    "print(\"\\n\\tFerredoxin\")\n",
    "DF = add_col_with_sequence(DF,'protein','ferredoxin')\n",
    "DF = add_col_with_sequence(DF,'CDS','ferredoxin')\n",
    "print(\"\\n\\tFlotillin\")\n",
    "DF = add_col_with_sequence(DF,'protein','flotillin')\n",
    "DF = add_col_with_sequence(DF,'CDS','flotillin')\n",
    "\n",
    "\n",
    "\n",
    "print(\"\\n#### V.    Loading start/end coordinates for rhodopsin genes\\n\")\n",
    "\n",
    "# Drop columns that will be made redundant with merge\n",
    "_DF = DF.drop(columns=['Unnamed: 0','sag','rhodopsin_scaffold'])\n",
    "\n",
    "_DF_annot = DF_annot[['rhodopsin_gene','fasta','scaffold','start_position','end_position','strandedness']]\n",
    "_DF_annot = _DF_annot.rename(columns={'fasta':'sag','scaffold':'rhodopsin_scaffold','start_position':'rhodopsin_start_position(DRAM)','end_position':'rhodopsin_end_position(DRAM)','strandedness':'rhodopsin_strandedness(DRAM)'})\n",
    "_DF_annot['sag'] = _DF_annot['sag'].str.replace('_contigs','')\n",
    "\n",
    "# Merge\n",
    "DF = _DF_annot.merge(_DF,on='rhodopsin_gene', how='right')\n",
    "\n",
    "\n",
    "print(\"\\n### VI.    Add operon info to table\")\n",
    "\n",
    "_DF_operons = DF_operons.drop(columns=['Unnamed: 0', '8th helix'])\n",
    "DF = DF.merge(_DF_operons,on='rhodopsin_gene',how='inner')\n",
    "\n",
    "\n",
    "\n",
    "print(\"\\n### VII.   Polishing and saving\")\n",
    "\n",
    "# Rename some columns for clarity\n",
    "DF = DF.rename(columns={'carotenoid_gene_cluster':'carotenoid_gene_cluster(HMMER3)',\n",
    "                       'flotillin':'flotillin(HMMER3)',\n",
    "                       'neither':'neither(HMMER3)',\n",
    "                       'flanking_genes':'flanking_genes(DRAM+HMMER+Foldseek)',\n",
    "                        'niche':'niched_based_on_read_recruitment',\n",
    "                      })\n",
    "\n",
    "print(\"\\tSaving...\")\n",
    "DF.to_csv(PATH_out,sep=\"\\t\")\n",
    "print(\"\\tDone.\")\n",
    "\n",
    "\n",
    "print(\"\\n### VIII.   Adding in the retinal_binding info that Tianyi forgot!\")\n",
    "\n",
    "DF_retinal = pd.read_csv(PATH_retinal_CSV)\n",
    "DF_retinal = DF_retinal[['seq_id','retinal_binding_residue','has_retinal_binding']]\n",
    "DF_retinal = DF_retinal.rename(columns={'seq_id':'rhodopsin_gene'})\n",
    "\n",
    "DF_main = pd.read_csv(PATH_out,sep=\"\\t\")\n",
    "\n",
    "LIST_col_order = [\"rhodopsin_gene\", 'sag', 'rhodopsin_scaffold',\n",
    "       'rhodopsin_start_position(DRAM)', 'rhodopsin_end_position(DRAM)',\n",
    "       'rhodopsin_strandedness(DRAM)', 'rhodopsin_count_in_sag',\n",
    "       'carotenoid_gene_cluster(HMMER3)', 'flotillin(HMMER3)',\n",
    "       'neither(HMMER3)', 'flanking_genes(DRAM+HMMER+Foldseek)',\n",
    "       'retinal_binding_residue','has_retinal_binding',\n",
    "       'color_tuning_type', 'color_tuning_residue', 'three_residue_motif',\n",
    "       'pump_type', 'gtdb', 'niched_based_on_read_recruitment',\n",
    "       'rhodopsin_family', 'rhodopsin_subfamily', 'suitable_for_phylogenetics',\n",
    "       '8th_N-terminal_helix (alphafold3)',\n",
    "       'incomplete_8th_helix (alphafold3)', 'rhodopsin_protein(DRAM)',\n",
    "       'rhodopsin_CDS(DRAM)', 'flanking_ferredoxin_protein(DRAM)',\n",
    "       'flanking_ferredoxin_CDS(DRAM)', 'flanking_flotillin_protein(DRAM)',\n",
    "       'flanking_flotillin_CDS(DRAM)', \"3' 2Fe-2S binding dom.\",\n",
    "       \"3' flotillin\", \"5' AMP-binding enz.\", \"5' VirC1\",\n",
    "       \"5' amidohydrolase fam.\", \"5' A.A. permease\",\n",
    "       \"5' periplasmic-binding prot.\", \"3' CrtE\", \"3' L-rhamnose isomerase\",\n",
    "       \"5' brp/blh\", \"3' tRNA synthetase I\", \"3' haemolysin-III related\",\n",
    "       \"5' fasciclin\", \"3' fatty acid desaturase\", \"3' brp/blh\",\n",
    "       \"5' thiamine pyrophosphate\", \"5' TMEM205-like\",\n",
    "       \"5' AI-2E fam. transporter\", \"5' toxin-antitoxin\",\n",
    "       \"3' phosphotransferase fam.\", \"3' SOUL heme-binding prot.\",\n",
    "       \"3' ABC1 kinase-like\", \"5' FGGY fam. carb. kinases\"]\n",
    "\n",
    "\n",
    "print(\"adding 2 columns\")\n",
    "DF = DF_main.merge(DF_retinal, on=\"rhodopsin_gene\", how=\"left\")\n",
    "DF = DF.drop(columns=[\"Unnamed: 0\"])\n",
    "DF = DF[LIST_col_order]\n",
    "\n",
    "print(\"Writing out updated table\")\n",
    "DF.to_csv(PATH_out, sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>rhodopsin_gene</th>\n",
       "      <th>fasta</th>\n",
       "      <th>scaffold</th>\n",
       "      <th>gene_position</th>\n",
       "      <th>start_position</th>\n",
       "      <th>end_position</th>\n",
       "      <th>strandedness</th>\n",
       "      <th>rank</th>\n",
       "      <th>ko_id</th>\n",
       "      <th>...</th>\n",
       "      <th>peptidase_RBH</th>\n",
       "      <th>peptidase_identity</th>\n",
       "      <th>peptidase_bitScore</th>\n",
       "      <th>peptidase_eVal</th>\n",
       "      <th>pfam_hits</th>\n",
       "      <th>cazy_ids</th>\n",
       "      <th>cazy_hits</th>\n",
       "      <th>cazy_subfam_ec</th>\n",
       "      <th>cazy_best_hit</th>\n",
       "      <th>heme_regulatory_motif_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>314</th>\n",
       "      <td>314</td>\n",
       "      <td>AG-538-A02_contigs_AG-538-A02_NODE_4_1</td>\n",
       "      <td>AG-538-A02_contigs</td>\n",
       "      <td>AG-538-A02_NODE_4</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>1466</td>\n",
       "      <td>-1</td>\n",
       "      <td>D</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Sodium:solute symporter family [PF00474.20]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>315</th>\n",
       "      <td>315</td>\n",
       "      <td>AG-538-A02_contigs_AG-538-A02_NODE_4_2</td>\n",
       "      <td>AG-538-A02_contigs</td>\n",
       "      <td>AG-538-A02_NODE_4</td>\n",
       "      <td>2</td>\n",
       "      <td>1471</td>\n",
       "      <td>1749</td>\n",
       "      <td>-1</td>\n",
       "      <td>C</td>\n",
       "      <td>K21700</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>316</th>\n",
       "      <td>316</td>\n",
       "      <td>AG-538-A02_contigs_AG-538-A02_NODE_4_3</td>\n",
       "      <td>AG-538-A02_contigs</td>\n",
       "      <td>AG-538-A02_NODE_4</td>\n",
       "      <td>3</td>\n",
       "      <td>1773</td>\n",
       "      <td>2141</td>\n",
       "      <td>-1</td>\n",
       "      <td>E</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>317</th>\n",
       "      <td>317</td>\n",
       "      <td>AG-538-A02_contigs_AG-538-A02_NODE_4_4</td>\n",
       "      <td>AG-538-A02_contigs</td>\n",
       "      <td>AG-538-A02_NODE_4</td>\n",
       "      <td>4</td>\n",
       "      <td>2141</td>\n",
       "      <td>4498</td>\n",
       "      <td>-1</td>\n",
       "      <td>D</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Domain of unknown function (DUF6605) [PF20254.1]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>318</th>\n",
       "      <td>318</td>\n",
       "      <td>AG-538-A02_contigs_AG-538-A02_NODE_4_5</td>\n",
       "      <td>AG-538-A02_contigs</td>\n",
       "      <td>AG-538-A02_NODE_4</td>\n",
       "      <td>5</td>\n",
       "      <td>4524</td>\n",
       "      <td>4931</td>\n",
       "      <td>-1</td>\n",
       "      <td>E</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>757039</th>\n",
       "      <td>435</td>\n",
       "      <td>AM-689-M16_contigs_SCGC_AM-689-M16_contig19_18</td>\n",
       "      <td>AM-689-M16_contigs</td>\n",
       "      <td>SCGC_AM-689-M16_contig19</td>\n",
       "      <td>18</td>\n",
       "      <td>16200</td>\n",
       "      <td>17348</td>\n",
       "      <td>1</td>\n",
       "      <td>E</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>757040</th>\n",
       "      <td>436</td>\n",
       "      <td>AM-689-M16_contigs_SCGC_AM-689-M16_contig19_19</td>\n",
       "      <td>AM-689-M16_contigs</td>\n",
       "      <td>SCGC_AM-689-M16_contig19</td>\n",
       "      <td>19</td>\n",
       "      <td>17356</td>\n",
       "      <td>17619</td>\n",
       "      <td>1</td>\n",
       "      <td>E</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>757041</th>\n",
       "      <td>437</td>\n",
       "      <td>AM-689-M16_contigs_SCGC_AM-689-M16_contig19_20</td>\n",
       "      <td>AM-689-M16_contigs</td>\n",
       "      <td>SCGC_AM-689-M16_contig19</td>\n",
       "      <td>20</td>\n",
       "      <td>17680</td>\n",
       "      <td>18339</td>\n",
       "      <td>-1</td>\n",
       "      <td>C</td>\n",
       "      <td>K07056</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Tetrapyrrole (Corrin/Porphyrin) Methylases [PF...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>757042</th>\n",
       "      <td>438</td>\n",
       "      <td>AM-689-M16_contigs_SCGC_AM-689-M16_contig19_21</td>\n",
       "      <td>AM-689-M16_contigs</td>\n",
       "      <td>SCGC_AM-689-M16_contig19</td>\n",
       "      <td>21</td>\n",
       "      <td>18411</td>\n",
       "      <td>19577</td>\n",
       "      <td>-1</td>\n",
       "      <td>C</td>\n",
       "      <td>K06194</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>0.545</td>\n",
       "      <td>130.0</td>\n",
       "      <td>1.527000e-30</td>\n",
       "      <td>Peptidase family M23 [PF01551.25]; LysM domain...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>757043</th>\n",
       "      <td>439</td>\n",
       "      <td>AM-689-M16_contigs_SCGC_AM-689-M16_contig19_22</td>\n",
       "      <td>AM-689-M16_contigs</td>\n",
       "      <td>SCGC_AM-689-M16_contig19</td>\n",
       "      <td>22</td>\n",
       "      <td>19772</td>\n",
       "      <td>19999</td>\n",
       "      <td>-1</td>\n",
       "      <td>E</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>39133 rows Ã— 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Unnamed: 0                                  rhodopsin_gene  \\\n",
       "314            314          AG-538-A02_contigs_AG-538-A02_NODE_4_1   \n",
       "315            315          AG-538-A02_contigs_AG-538-A02_NODE_4_2   \n",
       "316            316          AG-538-A02_contigs_AG-538-A02_NODE_4_3   \n",
       "317            317          AG-538-A02_contigs_AG-538-A02_NODE_4_4   \n",
       "318            318          AG-538-A02_contigs_AG-538-A02_NODE_4_5   \n",
       "...            ...                                             ...   \n",
       "757039         435  AM-689-M16_contigs_SCGC_AM-689-M16_contig19_18   \n",
       "757040         436  AM-689-M16_contigs_SCGC_AM-689-M16_contig19_19   \n",
       "757041         437  AM-689-M16_contigs_SCGC_AM-689-M16_contig19_20   \n",
       "757042         438  AM-689-M16_contigs_SCGC_AM-689-M16_contig19_21   \n",
       "757043         439  AM-689-M16_contigs_SCGC_AM-689-M16_contig19_22   \n",
       "\n",
       "                     fasta                  scaffold  gene_position  \\\n",
       "314     AG-538-A02_contigs         AG-538-A02_NODE_4              1   \n",
       "315     AG-538-A02_contigs         AG-538-A02_NODE_4              2   \n",
       "316     AG-538-A02_contigs         AG-538-A02_NODE_4              3   \n",
       "317     AG-538-A02_contigs         AG-538-A02_NODE_4              4   \n",
       "318     AG-538-A02_contigs         AG-538-A02_NODE_4              5   \n",
       "...                    ...                       ...            ...   \n",
       "757039  AM-689-M16_contigs  SCGC_AM-689-M16_contig19             18   \n",
       "757040  AM-689-M16_contigs  SCGC_AM-689-M16_contig19             19   \n",
       "757041  AM-689-M16_contigs  SCGC_AM-689-M16_contig19             20   \n",
       "757042  AM-689-M16_contigs  SCGC_AM-689-M16_contig19             21   \n",
       "757043  AM-689-M16_contigs  SCGC_AM-689-M16_contig19             22   \n",
       "\n",
       "        start_position  end_position  strandedness rank   ko_id  ...  \\\n",
       "314                  3          1466            -1    D     NaN  ...   \n",
       "315               1471          1749            -1    C  K21700  ...   \n",
       "316               1773          2141            -1    E     NaN  ...   \n",
       "317               2141          4498            -1    D     NaN  ...   \n",
       "318               4524          4931            -1    E     NaN  ...   \n",
       "...                ...           ...           ...  ...     ...  ...   \n",
       "757039           16200         17348             1    E     NaN  ...   \n",
       "757040           17356         17619             1    E     NaN  ...   \n",
       "757041           17680         18339            -1    C  K07056  ...   \n",
       "757042           18411         19577            -1    C  K06194  ...   \n",
       "757043           19772         19999            -1    E     NaN  ...   \n",
       "\n",
       "       peptidase_RBH peptidase_identity peptidase_bitScore peptidase_eVal  \\\n",
       "314              NaN                NaN                NaN            NaN   \n",
       "315              NaN                NaN                NaN            NaN   \n",
       "316              NaN                NaN                NaN            NaN   \n",
       "317              NaN                NaN                NaN            NaN   \n",
       "318              NaN                NaN                NaN            NaN   \n",
       "...              ...                ...                ...            ...   \n",
       "757039           NaN                NaN                NaN            NaN   \n",
       "757040           NaN                NaN                NaN            NaN   \n",
       "757041           NaN                NaN                NaN            NaN   \n",
       "757042         False              0.545              130.0   1.527000e-30   \n",
       "757043           NaN                NaN                NaN            NaN   \n",
       "\n",
       "                                                pfam_hits  cazy_ids  \\\n",
       "314           Sodium:solute symporter family [PF00474.20]       NaN   \n",
       "315                                                   NaN       NaN   \n",
       "316                                                   NaN       NaN   \n",
       "317      Domain of unknown function (DUF6605) [PF20254.1]       NaN   \n",
       "318                                                   NaN       NaN   \n",
       "...                                                   ...       ...   \n",
       "757039                                                NaN       NaN   \n",
       "757040                                                NaN       NaN   \n",
       "757041  Tetrapyrrole (Corrin/Porphyrin) Methylases [PF...       NaN   \n",
       "757042  Peptidase family M23 [PF01551.25]; LysM domain...       NaN   \n",
       "757043                                                NaN       NaN   \n",
       "\n",
       "        cazy_hits  cazy_subfam_ec cazy_best_hit heme_regulatory_motif_count  \n",
       "314           NaN             NaN           NaN                           0  \n",
       "315           NaN             NaN           NaN                           0  \n",
       "316           NaN             NaN           NaN                           0  \n",
       "317           NaN             NaN           NaN                           0  \n",
       "318           NaN             NaN           NaN                           0  \n",
       "...           ...             ...           ...                         ...  \n",
       "757039        NaN             NaN           NaN                           0  \n",
       "757040        NaN             NaN           NaN                           0  \n",
       "757041        NaN             NaN           NaN                           0  \n",
       "757042        NaN             NaN           NaN                           0  \n",
       "757043        NaN             NaN           NaN                           0  \n",
       "\n",
       "[39133 rows x 24 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DF_annot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['rhodopsin_gene', 'sag', 'rhodopsin_scaffold',\n",
       "       'rhodopsin_start_position(DRAM)', 'rhodopsin_end_position(DRAM)',\n",
       "       'rhodopsin_strandedness(DRAM)', 'rhodopsin_count_in_sag',\n",
       "       'carotenoid_gene_cluster(HMMER3)', 'flotillin(HMMER3)',\n",
       "       'neither(HMMER3)', 'flanking_genes(DRAM+HMMER+Foldseek)',\n",
       "       'color_tuning_type', 'color_tuning_residue', 'three_residue_motif',\n",
       "       'pump_type', 'gtdb', 'niched_based_on_read_recruitment',\n",
       "       'rhodopsin_family', 'rhodopsin_subfamily', 'suitable_for_phylogenetics',\n",
       "       '8th_N-terminal_helix (alphafold3)',\n",
       "       'incomplete_8th_helix (alphafold3)', 'rhodopsin_protein(DRAM)',\n",
       "       'rhodopsin_CDS(DRAM)', 'flanking_ferredoxin_protein(DRAM)',\n",
       "       'flanking_ferredoxin_CDS(DRAM)', 'flanking_flotillin_protein(DRAM)',\n",
       "       'flanking_flotillin_CDS(DRAM)', '3' 2Fe-2S binding dom.',\n",
       "       '3' flotillin', '5' AMP-binding enz.', '5' VirC1',\n",
       "       '5' amidohydrolase fam.', '5' A.A. permease',\n",
       "       '5' periplasmic-binding prot.', '3' CrtE', '3' L-rhamnose isomerase',\n",
       "       '5' brp/blh', '3' tRNA synthetase I', '3' haemolysin-III related',\n",
       "       '5' fasciclin', '3' fatty acid desaturase', '3' brp/blh',\n",
       "       '5' thiamine pyrophosphate', '5' TMEM205-like',\n",
       "       '5' AI-2E fam. transporter', '5' toxin-antitoxin',\n",
       "       '3' phosphotransferase fam.', '3' SOUL heme-binding prot.',\n",
       "       '3' ABC1 kinase-like', '5' FGGY fam. carb. kinases'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DF.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
