{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 5. Reannotate rhodopsin neighborhood\n",
    "\n",
    "We'll define rhodopsin's \"neighborhood\" as genes up to five positions downstrand (1...5 at furthest) and five positions upstream (-1...-5 at furthest) of rhodopsin.\n",
    "I.e. so we're looking at a more limited search window of gene annotations. (Only in the vicinity of rhodopsin, rather than the entire contig).\n",
    "This notebook has two steps:\n",
    "* (5a). Integrating all the annotations (from Dram, HMMsearch, and Foldseek) into a single table (output_A)\n",
    "* (5b). Transforming that into a column called 'flanking genes' that takes neighborhood info into account\n",
    "\n",
    "### 5a:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/mnt/scgc/scgc_nfs/opt/common/anaconda3a/lib/python3.7/site-packages/pandas/compat/_optional.py:138: UserWarning: Pandas requires version '2.7.0' or newer of 'numexpr' (version '2.6.9' currently installed).\n",
      "  warnings.warn(msg, UserWarning)\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "from pathlib import Path\n",
    "import re\n",
    "\n",
    "#### Output \n",
    "PATH_output_A = './output/A_dram_reannot_with_hmmscan_and_foldseek.csv'\n",
    "PATH_output_B = \"./output/B_rhodopsin_neighborhood.csv\"\n",
    "\n",
    "#### Inputs\n",
    "DIR_hmmscan = '../2_tabulate_hmmscan_hits/output/'\n",
    "PATH_foldseek = '../4_reannotate_dramAAs_based_on_foldseek_annot/rankE_reannotations.csv'\n",
    "PATH_dram = \"../0_get_DRAM_AAs_for_SAGs_w_rhodopsins/dram_926_SAGs_combined.csv\"\n",
    "\n",
    "####### I. ##### Combine annots from DRAM, hmmscan, & Foldseek into 1 table  -> (DF3) ######\n",
    "\n",
    "# Import the hmmscan annotation tables\n",
    "# Get all files matching the pattern '_hmmsearch_minbit10.csv'\n",
    "LIST_hmmscan_files = list(Path(DIR_hmmscan).glob('*_hmmsearch_minbit10.csv'))\n",
    "\n",
    "# Read and combine all matching files\n",
    "LIST_hmmscan_DFs = [pd.read_csv(file) for file in LIST_hmmscan_files]\n",
    "DF_hmmscan = pd.concat(LIST_hmmscan_DFs, ignore_index=True)\n",
    "\n",
    "# Select and rename columns\n",
    "DF_hmmscan = DF_hmmscan.rename(columns={\n",
    "    'query_name': 'cds_id',\n",
    "    'hmmsearching_for': 'hmmsearch_gene',\n",
    "    'score': 'hmmsearch_score',\n",
    "    'target_accession': 'hmmsearch_acc'\n",
    "})[['cds_id', 'hmmsearch_gene', 'hmmsearch_score', 'hmmsearch_acc']]\n",
    "\n",
    "\n",
    "# Remove prefix from cds_id (everything up to and including first underscore)\n",
    "DF_hmmscan['cds_id'] = DF_hmmscan['cds_id'].str.replace(r'^[^_]+_', '', regex=True)\n",
    "\n",
    "# Import the foldseek annotation table\n",
    "DF_foldseek = pd.read_csv(PATH_foldseek)[['gene_id', 'description']].rename(\n",
    "    columns={'description': 'foldseek_hit'}\n",
    ")\n",
    "\n",
    "# Import dram annotation table and perform joins\n",
    "DF_dram = pd.read_csv(PATH_dram)\n",
    "\n",
    "# Left join with DF_hmmscan\n",
    "DF2 = DF_dram.merge(DF_hmmscan, left_on='...1', right_on='cds_id', how='left')\n",
    "\n",
    "# Create gene_id by removing prefix up to '_contigs_'\n",
    "DF2['gene_id'] = DF2['...1'].str.replace(r'^.*_contigs_', '', regex=True)\n",
    "\n",
    "# Left join with foldseek\n",
    "DF3 = DF2.merge(DF_foldseek, on='gene_id', how='left')\n",
    "\n",
    "# Create annot_multisrc column: use foldseek_hit if available, otherwise pfam_hits\n",
    "DF3['annot_multisrc'] = DF3['foldseek_hit'].fillna(DF3['pfam_hits'])\n",
    "\n",
    "# Drop gene_id column\n",
    "DF3 = DF3.drop(columns=['gene_id'])\n",
    "\n",
    "# Write to CSV\n",
    "DF3.to_csv(PATH_output, index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 5b:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "####### II. ##### Get the rhodopsin neighborhood (i.e. annotations for 5 genes upstrand and 5 genes downstrand of rhodopsin)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "> python **5b_rhodopsin_neighborhood.py** --input_file \"./output/A_reannot_with_hmmscan_and_foldseek.csv\" --output=\"./output/B_rhodopsin_neighborhood\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Understanding 'flanking_genes' format..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>rhodopsin_gene</th>\n",
       "      <th>rhodopsin_scaffold</th>\n",
       "      <th>sag</th>\n",
       "      <th>rhodopsin_count_in_sag</th>\n",
       "      <th>carotenoid_gene_cluster</th>\n",
       "      <th>flotillin</th>\n",
       "      <th>neither</th>\n",
       "      <th>flanking_genes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AG-538-J23_contigs_AG-538-J23_NODE_9_13</td>\n",
       "      <td>AG-538-J23_NODE_9</td>\n",
       "      <td>AG-538-J23_contigs</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>{'annot_multisrc of gene -5': ['Glycosyl trans...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AG-538-C05_contigs_AG-538-C05_NODE_1_19</td>\n",
       "      <td>AG-538-C05_NODE_1</td>\n",
       "      <td>AG-538-C05_contigs</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>{'annot_multisrc of gene -5': ['Periplasmic bi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AG-538-N13_contigs_AG-538-N13_NODE_14_10</td>\n",
       "      <td>AG-538-N13_NODE_14</td>\n",
       "      <td>AG-538-N13_contigs</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>{'annot_multisrc of gene -5': ['Domain of unkn...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AH-302-C16_contigs_AH-302-C16_NODE_2_56</td>\n",
       "      <td>AH-302-C16_NODE_2</td>\n",
       "      <td>AH-302-C16_contigs</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>{'annot_multisrc of gene -5': ['Dolichyl-phosp...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AH-302-J23_contigs_AH-302-J23_NODE_61_2</td>\n",
       "      <td>AH-302-J23_NODE_61</td>\n",
       "      <td>AH-302-J23_contigs</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>{'annot_multisrc of gene -1': ['Domain of unkn...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                             rhodopsin_gene  rhodopsin_scaffold  \\\n",
       "0   AG-538-J23_contigs_AG-538-J23_NODE_9_13   AG-538-J23_NODE_9   \n",
       "1   AG-538-C05_contigs_AG-538-C05_NODE_1_19   AG-538-C05_NODE_1   \n",
       "2  AG-538-N13_contigs_AG-538-N13_NODE_14_10  AG-538-N13_NODE_14   \n",
       "3   AH-302-C16_contigs_AH-302-C16_NODE_2_56   AH-302-C16_NODE_2   \n",
       "4   AH-302-J23_contigs_AH-302-J23_NODE_61_2  AH-302-J23_NODE_61   \n",
       "\n",
       "                  sag  rhodopsin_count_in_sag  carotenoid_gene_cluster  \\\n",
       "0  AG-538-J23_contigs                       1                        0   \n",
       "1  AG-538-C05_contigs                       1                        0   \n",
       "2  AG-538-N13_contigs                       1                        0   \n",
       "3  AH-302-C16_contigs                       1                        0   \n",
       "4  AH-302-J23_contigs                       1                        0   \n",
       "\n",
       "   flotillin  neither                                     flanking_genes  \n",
       "0          0        0  {'annot_multisrc of gene -5': ['Glycosyl trans...  \n",
       "1          1        0  {'annot_multisrc of gene -5': ['Periplasmic bi...  \n",
       "2          1        0  {'annot_multisrc of gene -5': ['Domain of unkn...  \n",
       "3          0        0  {'annot_multisrc of gene -5': ['Dolichyl-phosp...  \n",
       "4          0        0  {'annot_multisrc of gene -1': ['Domain of unkn...  "
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DF_neighborhood = pd.read_csv(PATH_output_B)\n",
    "DF_neighborhood.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Let's look at an example of this 'Flanking_genes' column..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\"{'annot_multisrc of gene -5': ['Glycosyl transferase family 2 [PF00535.29]', 'None', 8230, 9054, 1], 'annot_multisrc of gene -4': ['Glucose / Sorbosone dehydrogenase [PF07995.14]', 'None', 9062, 10432, 1], 'annot_multisrc of gene -3': ['GDP-mannose 4,6 dehydratase [PF16363.8]; NAD dependent epimerase/dehydratase family [PF01370.24]; 3-beta hydroxysteroid dehydrogenase/isomerase family [PF01073.22]; RmlD substrate binding domain [PF04321.20]; Polysaccharide biosynthesis protein [PF02719.18]; Male sterility protein [PF07993.15]; NmrA-like family [PF05368.16]', 'None', 10366, 11406, -1], 'annot_multisrc of gene -2': ['Small Multidrug Resistance protein [PF00893.22]', 'None', 11410, 11742, -1], 'annot_multisrc of gene -1': ['AMP-binding enzyme [PF00501.31]; Acetyl-coenzyme A synthetase N-terminus [PF16177.8]; AMP-binding enzyme C-terminal domain [PF13193.9]', 'None', 11742, 13637, -1], 'annot_multisrc of rhodopsin gene': ['Bacteriorhodopsin-like protein [PF01036.21]', 'None', 13834, 14601, 1], 'annot_multisrc of gene +1': ['2Fe-2S iron-sulfur cluster binding domain [PF00111.30]', 'None', 14680, 15000, 1], 'annot_multisrc of gene +2': ['Pyridine nucleotide-disulphide oxidoreductase [PF07992.17]; Flavin-binding monooxygenase-like [PF00743.22]; HI0933-like protein [PF03486.17]; NAD(P)-binding Rossmann-like domain [PF13450.9]', 'CrtI', 15000, 16022, 1], 'annot_multisrc of gene +3': ['Glutathione S-transferase, N-terminal domain [PF02798.23]; Glutathione S-transferase, C-terminal domain [PF00043.28]; Glutathione S-transferase, N-terminal domain [PF13409.9]; Glutathione S-transferase, N-terminal domain [PF13417.9]; Glutathione S-transferase, C-terminal domain [PF14497.9]', 'None', 16015, 16662, 1], 'annot_multisrc of gene +4': ['Prokaryotic dksA/traR C4-type zinc finger [PF01258.20]', 'None', 16659, 17075, -1], 'annot_multisrc of gene +5': ['Glycosyltransferase RgtA/B/C/D-like domain-containing protein', 'None', 17154, 17408, 1]}\""
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DF_neighborhood['flanking_genes'][0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'annot_multisrc of gene -5': ['Glycosyl transferase family 2 [PF00535.29]',\n",
       "  'None',\n",
       "  8230,\n",
       "  9054,\n",
       "  1],\n",
       " 'annot_multisrc of gene -4': ['Glucose / Sorbosone dehydrogenase [PF07995.14]',\n",
       "  'None',\n",
       "  9062,\n",
       "  10432,\n",
       "  1],\n",
       " 'annot_multisrc of gene -3': ['GDP-mannose 4,6 dehydratase [PF16363.8]; NAD dependent epimerase/dehydratase family [PF01370.24]; 3-beta hydroxysteroid dehydrogenase/isomerase family [PF01073.22]; RmlD substrate binding domain [PF04321.20]; Polysaccharide biosynthesis protein [PF02719.18]; Male sterility protein [PF07993.15]; NmrA-like family [PF05368.16]',\n",
       "  'None',\n",
       "  10366,\n",
       "  11406,\n",
       "  -1],\n",
       " 'annot_multisrc of gene -2': ['Small Multidrug Resistance protein [PF00893.22]',\n",
       "  'None',\n",
       "  11410,\n",
       "  11742,\n",
       "  -1],\n",
       " 'annot_multisrc of gene -1': ['AMP-binding enzyme [PF00501.31]; Acetyl-coenzyme A synthetase N-terminus [PF16177.8]; AMP-binding enzyme C-terminal domain [PF13193.9]',\n",
       "  'None',\n",
       "  11742,\n",
       "  13637,\n",
       "  -1],\n",
       " 'annot_multisrc of rhodopsin gene': ['Bacteriorhodopsin-like protein [PF01036.21]',\n",
       "  'None',\n",
       "  13834,\n",
       "  14601,\n",
       "  1],\n",
       " 'annot_multisrc of gene +1': ['2Fe-2S iron-sulfur cluster binding domain [PF00111.30]',\n",
       "  'None',\n",
       "  14680,\n",
       "  15000,\n",
       "  1],\n",
       " 'annot_multisrc of gene +2': ['Pyridine nucleotide-disulphide oxidoreductase [PF07992.17]; Flavin-binding monooxygenase-like [PF00743.22]; HI0933-like protein [PF03486.17]; NAD(P)-binding Rossmann-like domain [PF13450.9]',\n",
       "  'CrtI',\n",
       "  15000,\n",
       "  16022,\n",
       "  1],\n",
       " 'annot_multisrc of gene +3': ['Glutathione S-transferase, N-terminal domain [PF02798.23]; Glutathione S-transferase, C-terminal domain [PF00043.28]; Glutathione S-transferase, N-terminal domain [PF13409.9]; Glutathione S-transferase, N-terminal domain [PF13417.9]; Glutathione S-transferase, C-terminal domain [PF14497.9]',\n",
       "  'None',\n",
       "  16015,\n",
       "  16662,\n",
       "  1],\n",
       " 'annot_multisrc of gene +4': ['Prokaryotic dksA/traR C4-type zinc finger [PF01258.20]',\n",
       "  'None',\n",
       "  16659,\n",
       "  17075,\n",
       "  -1],\n",
       " 'annot_multisrc of gene +5': ['Glycosyltransferase RgtA/B/C/D-like domain-containing protein',\n",
       "  'None',\n",
       "  17154,\n",
       "  17408,\n",
       "  1]}"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import ast\n",
    "DICT_neighborhood_of_rhodA = ast.literal_eval(DF_neighborhood['flanking_genes'][0])\n",
    "DICT_neighborhood_of_rhodA"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### So, for instance: This LIST describes the protein 2 steps downstream from rhodopsin1..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Pyridine nucleotide-disulphide oxidoreductase [PF07992.17]; Flavin-binding monooxygenase-like [PF00743.22]; HI0933-like protein [PF03486.17]; NAD(P)-binding Rossmann-like domain [PF13450.9]',\n",
       " 'CrtI',\n",
       " 15000,\n",
       " 16022,\n",
       " 1]"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DICT_neighborhood_of_rhodA['annot_multisrc of gene +2']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Where LIST[0] is annots from DRAM or Foldseek (if no DRAM annot was found):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Pyridine nucleotide-disulphide oxidoreductase [PF07992.17]; Flavin-binding monooxygenase-like [PF00743.22]; HI0933-like protein [PF03486.17]; NAD(P)-binding Rossmann-like domain [PF13450.9]'"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DICT_neighborhood_of_rhodA['annot_multisrc of gene +2'][0]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### LIST[1] is the annots from HMMsearch:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'CrtI'"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DICT_neighborhood_of_rhodA['annot_multisrc of gene +2'][1]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### LIST[2,3] are the starting and ending nuc coords for that protein:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[15000, 16022]"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DICT_neighborhood_of_rhodA['annot_multisrc of gene +2'][2:4]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### And LIST[4] is the strandedness (1 for pos, -1 for neg):\n",
    "DICT_neighborhood_of_rhodA['annot_multisrc of gene +2'][4]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DICT_neighborhood_of_rhodA['annot_multisrc of gene +2'][4]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
