{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 6. Most common operons\n",
    "\n",
    "Now that we've richly annotated 5 the 5-gene radius surrounding rhodopsins, we want to know...<br>\n",
    "What annotations are the most common in these flanking genes?<br>\n",
    "In order to uncover operons or gene blocks<Br>\n",
    "Which could illuminate the functional context of each rhodopsin"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Getting strandedness of rhodopsin from the massive Dram table\n",
      "done\n",
      "Loading gene neighborhood table to get flanking gene info\n",
      "Splitting flanking gene info into 10 columns.\n",
      "\n",
      "Of 1095 contigs with rhodopsin...\n",
      "572 rhodopsins are on the forward/positive strand\n",
      "523 rhodopsins are on the reverse/negative strand\n",
      "For those negative-stranded rhodopsins, renumbered the flanking genes so that neg positions become positive and vice versa\n",
      "In other words, flanking genes are numbered -5 to +5 with respect to the 5'->3' orientation or rhodopsin.\n",
      "1095 contigs following re-orientation\n",
      "\n",
      "\n",
      "~~~~~~~ Annotations Flanking Downstream ~~~~~~~~~\n",
      "\n",
      "3' 2Fe-2S binding dom. ~ PF00111 ~ 422\n",
      "3' flotillin ~ lotillin ~ 391\n",
      "3' CrtE ~ CrtE ~ 25\n",
      "3' L-rhamnose isomerase ~ PF06134 ~ 24\n",
      "3' fatty acid desaturase ~ PF00487 ~ 9\n",
      "3' brp/blh ~ blh ~ 8\n",
      "3' tRNA synthetase I ~ PF00750 ~ 9\n",
      "3' phosphotransferase fam. ~ PF01636 ~ 6\n",
      "3' SOUL heme-binding prot. ~ PF04832 ~ 6\n",
      "3' ABC1 kinase-like ~ PF03109 ~ 6\n",
      "3' haemolysin-III related ~ PF03006 ~ 9\n",
      "\n",
      "\n",
      "~~~~~~~~~ Annotations Flanking Upstream ~~~~~~~~~~~\n",
      "\n",
      "5' AMP-binding enz. ~ PF00501 ~ 293\n",
      "5' VirC1 ~ VirC1 ~ 233\n",
      "5' amidohydrolase fam. ~ PF01979 ~ 88\n",
      "5' A.A. permease ~ PF02653 ~ 59\n",
      "5' periplasmic-binding prot. ~ PF13407 ~ 43\n",
      "5' brp/blh ~ blh ~ 9\n",
      "5' fasciclin ~ PF02469 ~ 9\n",
      "5' thiamine pyrophosphate ~ PF02776 ~ 8\n",
      "5' TMEM205-like ~ TMEM205 ~ 8\n",
      "5' AI-2E fam. transporter ~ PF01594 ~ 8\n",
      "5' toxin-antitoxin ~ toxin ~ 7\n",
      "5' FGGY fam. carb. kinases ~ PF00370 ~ 5\n",
      "\n",
      "Saved results to Presence/Absence table.\n",
      "Use the table to the plot presence/absence of flanking genes on iToL phylogeny\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from os.path import exists\n",
    "import ast\n",
    "import sys\n",
    "import itertools\n",
    "import collections\n",
    "from Bio import SeqIO\n",
    "\n",
    "MODE_verbose = False # Make 'True' to see report in Part I\n",
    "MODE_printIDS = False  #Super-verbose: Prints the IDs of each contig containing rhodopsin\n",
    "\n",
    "LIST_cols_most_to_least_common = [\"3' 2Fe-2S binding dom.\", \"3' flotillin\", \"5' AMP-binding enz.\", \"5' VirC1\", \"5' amidohydrolase fam.\", \"5' A.A. permease\", \"5' periplasmic-binding prot.\", \"3' CrtE\", \"3' L-rhamnose isomerase\", \"5' brp/blh\", \"3' tRNA synthetase I\", \"3' haemolysin-III related\", \"5' fasciclin\", \"3' fatty acid desaturase\", \"3' brp/blh\", \"5' thiamine pyrophosphate\", \"5' TMEM205-like\", \"5' AI-2E fam. transporter\", \"5' toxin-antitoxin\", \"3' phosphotransferase fam.\", \"3' SOUL heme-binding prot.\", \"3' ABC1 kinase-like\", \"5' FGGY fam. carb. kinases\"]\n",
    "\n",
    "PATH_out = \"./output/rhodopsin_operon_presence_absence_Dec11.csv\"\n",
    "\n",
    "PATH_in = \"../5_reannotate_rhodopsin_neighborhood/output/B_rhodopsin_neighborhood.csv\"\n",
    "#PATH_in = './input/rhodopsin_neighborhood.csv' # w/ foldseek annots\n",
    "\n",
    "# Need Dram table to get strand info for the rhodopsins\n",
    "PATH_dram = \"../0_get_DRAM_AAs_for_SAGs_w_rhodopsins/dram_926_SAGs_combined.csv\"\n",
    "#PATH_dram = '/mnt/scgc/stepanauskas_nfs/projects/gorg-dark/dram/dram_annot_251010.csv'\n",
    "\n",
    "PATH_temp = './temp/rhodopsin_neighborhood_plus_strandedness.csv' # Saves time re-loading Dram\n",
    "\n",
    "LIST_key_order = ['annot_multisrc of gene -5','annot_multisrc of gene -4','annot_multisrc of gene -3','annot_multisrc of gene -2','annot_multisrc of gene -1','annot_multisrc of gene +1','annot_multisrc of gene +2','annot_multisrc of gene +3','annot_multisrc of gene +4','annot_multisrc of gene +5']\n",
    "\n",
    "#### Top 20 annotations UPSTREAM of rhodopsin (at pos -1) and their frequency (manually tabulated)\n",
    "## Note that there are fewer than 20 because many annotations were variations of the same thing (i.e. reworded but sharing a Pfam)\n",
    "#1 AMP-binding enzyme [PF00501.31]; Acetyl-coenzyme A synthetase N-terminus [PF16177.8]; AMP-binding enzyme C-terminal domain [PF13193.9]   280\n",
    "#2 Unannotated    165 (Can't search for this)\n",
    "#3 \"VirC1\" protein 230\n",
    "#4 Amidohydrolase family [PF01979.23]; Amidohydrolase family [PF07969.14] 89\n",
    "#5 Branched-chain amino acid transport system / permease component [PF02653.19]   59\n",
    "#6 Periplasmic binding protein domain [PF13407.9]; Periplasmic binding proteins and sugar binding domain of LacI family [PF00532.24]  41\n",
    "#7 AMP-binding enzyme [PF00501.31] 11\n",
    "#8 Beta-carotene 15,15'-dioxygenase [PF15461.9]|brp/blh \n",
    "#9 Fasciclin domain [PF02469.25] \n",
    "#10 Thiamine pyrophosphate enzyme, N-terminal TPP binding domain [PF02776.21] 8\n",
    "#11 TMEM205-like domain-containing protein  8\n",
    "#12 AI-2E family transporter [PF01594.19]    8\n",
    "#13 SUKH-4 immunity protein of toxin-antitoxin system       6\n",
    "#14 FGGY family of carbohydrate kinases, N-terminal domain [PF00370.24]    5\n",
    " \n",
    "#### Distilled to their core unique patterns (Pfam if possible, but substring if pfam is lacking or if substring is more prevalent)\n",
    "DICT_upstream_annotPattern2Nickname={\n",
    "\"PF00501\":\"5' AMP-binding enz.\",\n",
    "\"VirC1\":\"5' VirC1\",\n",
    "\"PF01979\":\"5' amidohydrolase fam.\",\n",
    "\"PF02653\":\"5' A.A. permease\",\n",
    "\"PF13407\":\"5' periplasmic-binding prot.\",\n",
    "\"blh\":\"5' brp/blh\",\n",
    "\"PF02469\":\"5' fasciclin\",\n",
    "\"PF02776\":\"5' thiamine pyrophosphate\",\n",
    "\"TMEM205\":\"5' TMEM205-like\",\n",
    "\"PF01594\":\"5' AI-2E fam. transporter\",\n",
    "\"toxin\":\"5' toxin-antitoxin\",\n",
    "\"PF00370\":\"5' FGGY fam. carb. kinases\"}\n",
    "\n",
    "#### Top 20 patterns DOWNSTREAM (at pos +1) and their frequency (manually tabulated)\n",
    "## Note that there are fewer than 20 because many annotations were variations of teh same thing (i.e. sharing a Pfam)\n",
    "#1 2Fe-2S iron-sulfur cluster binding domain [PF00111.30]   422\n",
    "#2 flotillin    384\n",
    "#3 Unannotated    83 (Can't search for this)\n",
    "#4 Polyprenyl synthetase [PF00348.20]|CrtE    25\n",
    "#5 L-rhamnose isomerase (RhaA) [PF06134.14]   24\n",
    "#6 Fatty acid desaturase [PF00487.27] 9 \n",
    "#7  Beta-carotene 15,15'-dioxygenase [PF15461.9]|brp/blh   7\n",
    "#8 tRNA synthetases class I (R) [PF00750.22]; DALR anticodon binding domain [PF05746.18]   7\n",
    "#9 Phosphotransferase enzyme family [PF01636.26]; Choline/ethanolamine kinase [PF01633.23]; Ecdysteroid kinase-like family [PF02958.23]  6\n",
    "#10 SOUL heme-binding protein [PF04832.15]                                                                  6\n",
    "#11 ABC1 atypical kinase-like domain [PF03109.19]; Cytochrome P450 [PF00067.25]; Protein kinase domain [PF00069.28]                   6\n",
    "#12 Haemolysin-III related [PF03006.23]                           5\n",
    "\n",
    "#### Distilled to their core unique patterns (Pfam if possible, but substring if pfam is lacking or if substring is more prevalent)\n",
    "DICT_downstream_annotPattern2Nickname={\n",
    "\"PF00111\":\"3' 2Fe-2S binding dom.\",\n",
    "\"lotillin\":\"3' flotillin\",\n",
    "\"CrtE\":\"3' CrtE\",\n",
    "\"PF06134\":\"3' L-rhamnose isomerase\",\n",
    "\"PF00487\":\"3' fatty acid desaturase\",\n",
    "\"blh\":\"3' brp/blh\",\n",
    "\"PF00750\":\"3' tRNA synthetase I\",\n",
    "\"PF01636\":\"3' phosphotransferase fam.\",\n",
    "\"PF04832\":\"3' SOUL heme-binding prot.\",\n",
    "\"PF03109\":\"3' ABC1 kinase-like\",\n",
    "\"PF03006\":\"3' haemolysin-III related\"\n",
    "}\n",
    "\n",
    "\n",
    "def I_string_dict_to_dictionary(STR_of_dict):\n",
    "    '''\n",
    "    For parseability, we want each dictlike entry to have 10 keys like:\n",
    "    ['annot_multisrc of gene -5','annot_multisrc of gene -4','annot_multisrc of gene -3','annot_multisrc of gene -2','annot_multisrc of gene -1','annot_multisrc of gene +1','annot_multisrc of gene +2','annot_multisrc of gene +3','annot_multisrc of gene +4','annot_multisrc of gene +5']\n",
    "    But if the contig was too short, some keys are missing.\n",
    "    So create those missing keys...\n",
    "    and give them vals that are empty strings.\n",
    "    '''\n",
    "    # Add missing positions/keys if necessary\n",
    "    DICT_hits = ast.literal_eval(STR_of_dict)\n",
    "    for key in LIST_key_order:\n",
    "        if key not in DICT_hits.keys():\n",
    "            DICT_hits[key]=''\n",
    "            \n",
    "    # reorder dictionary so stays in LIST_key_order\n",
    "    orderedDICT_hits = collections.OrderedDict((k, DICT_hits[k]) for k in LIST_key_order)\n",
    "    DICT_hits = dict(orderedDICT_hits)\n",
    "    \n",
    "    return DICT_hits\n",
    "\n",
    "def II_dict_to_tilda_separated_string(DICT):\n",
    "    '''\n",
    "    '''\n",
    "    STR_from_dict = \"\"\n",
    "    for key in ['annot_multisrc of gene -5','annot_multisrc of gene -4','annot_multisrc of gene -3','annot_multisrc of gene -2','annot_multisrc of gene -1','annot_multisrc of gene +1','annot_multisrc of gene +2','annot_multisrc of gene +3','annot_multisrc of gene +4','annot_multisrc of gene +5']:\n",
    "        LIST_Pfam_hits = DICT[key]\n",
    "        #Drop Pfam hits that are just \"None\"\n",
    "        LIST_Pfam_hits = [i for i in LIST_Pfam_hits if i != 'None'] \n",
    "        #Drop Pfam hits that are just numbers\n",
    "        LIST_Pfam_hits = [i for i in LIST_Pfam_hits if type(i) == str]\n",
    "        if len(LIST_Pfam_hits) == 0:\n",
    "            LIST_Pfam_hits = ''\n",
    "        val = '|'.join(LIST_Pfam_hits)\n",
    "        STR_from_dict = STR_from_dict+val+\"~\"\n",
    "    STR_from_dict = STR_from_dict[:-1] # Drop trailing tilda\n",
    "    return STR_from_dict\n",
    "\n",
    "def III_list_annots_up_or_downstream_of_rhodopsin_starting_at_specific_pattern(DF_in, desired_substring_in_annotation, up_OR_down, MODE_annots_or_occupancy_or_rhodopsinIDs):\n",
    "    '''\n",
    "    If MODE_annots_or_occupancy_or_rhodopsinIDs == 'annots':\n",
    "    E.g. We've seen that \"flotillin\" is enriched near rhodopsin, so...\n",
    "    This function would...\n",
    "        Subset the DF so we're only looking at neighborhods where substring \"flotillin\" is found in the annotation next to rhodopsin\n",
    "        If set to upstream (\"up\"):\n",
    "            Requires \"flotillin\" at -1, then reruns a list of upstream annots (-5,-4,-3,-2,-1)\n",
    "        If set to downstream (\"down\"):\n",
    "            Requires \"flotillin\" at +1, then returns a list of downstream annots (+1,+2,+3,+4,+5)\n",
    "    \n",
    "    E.g. returns: ['flotillin', 'carboxyblahblah PF230303','','DUF224','actin PF24545']\n",
    "    \n",
    "    INSTEAD: if MODE_annots_or_occupancy_or_rhodopsinIDs == 'occupancy':\n",
    "    returns a list of the frequency of that annot at each site:\n",
    "    \n",
    "    E.g. returns [1.0, 0.09, 0.3, 0.1 0.0 ]\n",
    "    \n",
    "    INSTEAD: if MODE_annots_or_occupancy_or_rhodopsinIDs == 'rhodopsinIDs'\n",
    "    returns a list of all rhodopsins with that pattern down/upstream (i.e. all rhodopsins with that putative operon-type)\n",
    "    '''\n",
    "        \n",
    "    LIST_annots = []\n",
    "    LIST_occupancy = []\n",
    "    \n",
    "    if up_OR_down=='up':\n",
    "        LIST_coords = [\"-5\",\"-4\",\"-3\",\"-2\",\"-1\"]\n",
    "        _DF = DF_in.loc[DF_in[\"-1\"].str.contains(desired_substring_in_annotation)]\n",
    "        #LIST_annots=[_DF[i].value_counts()[0] for i in LIST_coords]\n",
    "        \n",
    "    elif up_OR_down=='down':\n",
    "        LIST_coords = [\"+1\",\"+2\",\"+3\",\"+4\",\"+5\"]\n",
    "        _DF = DF_in.loc[DF_in[\"+1\"].str.contains(desired_substring_in_annotation)]\n",
    "        #LIST_annots=[_DF[i].value_counts()[0] for i in LIST_coords]\n",
    "    else:\n",
    "        sys.exit(\"Error. You must specify 'up' or 'down'!\")\n",
    "    \n",
    "    LIST_rhodopsin_ids = _DF['rhodopsin_gene'].to_list()\n",
    "    \n",
    "    for pos in LIST_coords:\n",
    "        DICT_annots_at_this_position = _DF[pos].value_counts()\n",
    "        top_annot = list(DICT_annots_at_this_position.keys())[0]\n",
    "        \n",
    "        # How many times is that annotation pattern at that position (e.g. 'PF00501' at position +1)\n",
    "        num_this_annot = DICT_annots_at_this_position[top_annot]\n",
    "        # Occupancy rate = perc of proteins at that position with that pattern (e.g. 0.5, a.k.a. 50% at pos +1 contain 'PF00501')\n",
    "        rate_occupancy = round(num_this_annot / len(_DF),2)\n",
    "       \n",
    "        LIST_annots.append(top_annot)\n",
    "        LIST_occupancy.append(rate_occupancy)\n",
    "        \n",
    "    if MODE_annots_or_occupancy_or_rhodopsinIDs == 'annots':\n",
    "        return LIST_annots\n",
    "    elif MODE_annots_or_occupancy_or_rhodopsinIDs == 'occupancy':\n",
    "        return LIST_occupancy\n",
    "    elif MODE_annots_or_occupancy_or_rhodopsinIDs == 'rhodopsinIDs':\n",
    "        return LIST_rhodopsin_ids\n",
    "\n",
    "### Read in Tianyi's neighborhood table\n",
    "DF_raw = pd.read_csv(PATH_in)\n",
    "\n",
    "### Still needs strand info, for rhodopsins, so retrieve that from Dram\n",
    "if not exists(PATH_temp):\n",
    "    print(\"Getting strandedness of rhodopsin from the massive Dram table\")\n",
    "    DF_dram = pd.read_csv(PATH_dram)\n",
    "    DF_dram = DF_dram[['...1','strandedness']]\n",
    "    DF_dram = DF_dram.rename(columns={'...1':'rhodopsin_gene'})\n",
    "    DF_merged = DF_raw.merge(DF_dram, on='rhodopsin_gene', how='left')\n",
    "    DF_merged = DF_merged.rename(columns={'strandedness':'rhodopsin_strandedness'})\n",
    "    DF_merged.to_csv(PATH_temp, index=False)\n",
    "    print(\"done\")\n",
    "    \n",
    "print(\"Loading gene neighborhood table to get flanking gene info\")\n",
    "DF1 = pd.read_csv(PATH_temp, index_col=None)\n",
    "\n",
    "\n",
    "### Preformat 'flanking genes' column before I can split it into 10 columns \n",
    "DF1['TMP_dict'] = DF1['flanking_genes'].apply(I_string_dict_to_dictionary)\n",
    "DF1['TMP_tildasep'] = DF1['TMP_dict'].apply(II_dict_to_tilda_separated_string)\n",
    "\n",
    "print(\"Splitting flanking gene info into 10 columns.\")\n",
    "### Split 'flanking genes' into 10 gene neighborhood columns, by coordinate\n",
    "DF1[['-5','-4','-3','-2','-1','+1','+2','+3','+4','+5']] = DF1['TMP_tildasep'].str.split('~',expand=True)\n",
    "\n",
    "### Drop temp columns\n",
    "DF1 = DF1.drop(columns=['TMP_dict','TMP_tildasep'])\n",
    "\n",
    "### For rows where rhodopsin is on negative strand (-1), flip the order of the flanking genes\n",
    "# Requires splitting the table temporarily\n",
    "\n",
    "print('\\nOf '+str(len(DF1))+' contigs with rhodopsin...')\n",
    "DF1forward = DF1.loc[DF1['rhodopsin_strandedness']==1]\n",
    "print(str(len(DF1forward))+\" rhodopsins are on the forward/positive strand\")\n",
    "\n",
    "DF1reverse = DF1.loc[DF1['rhodopsin_strandedness']==-1]\n",
    "print(str(len(DF1reverse))+\" rhodopsins are on the reverse/negative strand\")\n",
    "\n",
    "DF1_rev_fixed = DF1reverse\n",
    "DF1_rev_fixed = DF1_rev_fixed.drop(columns=['-5','-4','-3','-2','-1','+1','+2','+3','+4','+5'])\n",
    "DF1_rev_fixed['-5'] = DF1reverse['+5']\n",
    "DF1_rev_fixed['-4'] = DF1reverse['+4']\n",
    "DF1_rev_fixed['-3'] = DF1reverse['+3']\n",
    "DF1_rev_fixed['-2'] = DF1reverse['+2']\n",
    "DF1_rev_fixed['-1'] = DF1reverse['+1']\n",
    "DF1_rev_fixed['+1'] = DF1reverse['-1']\n",
    "DF1_rev_fixed['+2'] = DF1reverse['-2']\n",
    "DF1_rev_fixed['+3'] = DF1reverse['-3']\n",
    "DF1_rev_fixed['+4'] = DF1reverse['-4']\n",
    "DF1_rev_fixed['+5'] = DF1reverse['-5']\n",
    "\n",
    "print(\"For those negative-stranded rhodopsins, renumbered the flanking genes so that neg positions become positive and vice versa\")\n",
    "print(\"In other words, flanking genes are numbered -5 to +5 with respect to the 5'->3' orientation or rhodopsin.\")\n",
    "\n",
    "DF2 = pd.concat([DF1forward, DF1_rev_fixed])\n",
    "DF2 = DF2.sort_values(by=['rhodopsin_gene'])\n",
    "print(str(len(DF2))+\" contigs following re-orientation\")\n",
    "\n",
    "\n",
    "########### PART I. DISPLAY MOST COMMON NEIGHBORING ANNOTATIONS (for exploratory purposes, i.e. to discover operons) ##################\n",
    "\n",
    "count_operonic_rhodopsins = 0\n",
    "\n",
    "if MODE_verbose==True:\n",
    "    print('\\nSearching for potential operons, by walking outward from most common neighbors to rhodopsin.')\n",
    "    \n",
    "    print(\"\\n\\nREPORTING UPSTREAM GENE NEIGHBORHOOD\")\n",
    "    count=0\n",
    "    #for up_pattern in LIST_search_upstream:\n",
    "    for up_pattern in list(DICT_upstream_annotPattern2Nickname.keys()):\n",
    "        count+=1\n",
    "        _LIST_annots = III_list_annots_up_or_downstream_of_rhodopsin_starting_at_specific_pattern(DF2, up_pattern, 'up', 'annots')\n",
    "\n",
    "        _LIST_occupancy_per_annot = III_list_annots_up_or_downstream_of_rhodopsin_starting_at_specific_pattern(DF2, up_pattern, 'up', 'occupancy')\n",
    "        _LIST_occupancy_per_annot = [str(i) for i in _LIST_occupancy_per_annot]\n",
    "    \n",
    "        # All these operons ended with rhodopsin\n",
    "        _LIST_annots.append('RHODOPSIN')\n",
    "        _LIST_occupancy_per_annot.append('1.0')\n",
    "    \n",
    "        count_operonic_rhodopsins = count_operonic_rhodopsins+len(DF2.loc[DF2['-1'].str.contains(up_pattern)])\n",
    "    \n",
    "        print(\"\\n\\n\"+str(len(DF2.loc[DF2['-1'].str.contains(up_pattern)]))+' of neighborhoods start from this enriched annotation/substring in the upstream neighbor of rhodopsin: \"'+ up_pattern+'\"')\n",
    "        print(\"with annotations in this order  (1 gene per line):\\n\" + '\\n*'.join(_LIST_annots))\n",
    "        print(\"\\nWhere X% of each position matches the annotations above:\\n\"+','.join(_LIST_occupancy_per_annot))\n",
    "\n",
    "        if MODE_printIDS==True:\n",
    "            print(\"\\nThe following rhodopsins have this neighboring annot:\\n\")\n",
    "            print(\",\".join(III_list_annots_up_or_downstream_of_rhodopsin_starting_at_specific_pattern(DF2, up_pattern, 'up', 'rhodopsinIDs')))\n",
    "        \n",
    "if MODE_verbose==True:\n",
    "    print(\"\\n\\nREPORTING DOWNSTREAM GENE NEIGHBORHOOD\")\n",
    "    count=0\n",
    "    #for down_pattern in LIST_search_downstream:\n",
    "    for down_pattern in list(DICT_downstream_annotPattern2Nickname.keys()):\n",
    "        count+=1\n",
    "        _LIST_annots = III_list_annots_up_or_downstream_of_rhodopsin_starting_at_specific_pattern(DF2, down_pattern, 'down', 'annots')\n",
    "    \n",
    "        _LIST_occupancy_per_annot = III_list_annots_up_or_downstream_of_rhodopsin_starting_at_specific_pattern(DF2, down_pattern, 'down', 'occupancy')\n",
    "        _LIST_occupancy_per_annot = [str(i) for i in _LIST_occupancy_per_annot]\n",
    "    \n",
    "        # All these operons started with rhodopsin\n",
    "        _LIST_annots = ['RHODOPSIN']+_LIST_annots\n",
    "        _LIST_occupancy_per_annot = ['1.0']+_LIST_occupancy_per_annot\n",
    "    \n",
    "        count_operonic_rhodopsins = count_operonic_rhodopsins+len(DF2.loc[DF2['-1'].str.contains(up_pattern)])\n",
    "    \n",
    "        print(\"\\n\\n\"+str(len(DF2.loc[DF2['+1'].str.contains(down_pattern)]))+' of neighborhoods start from this enriched annotation/substring in the downstream neighbor of rhodopsin: \"'+ down_pattern+'\"')\n",
    "        print(\"with annotations in this order (1 gene per line):\\n\" + '\\n*'.join(_LIST_annots))\n",
    "        print(\"\\n Where X% of each position matches the annotations above:\\n\"+','.join(_LIST_occupancy_per_annot))\n",
    "\n",
    "        if MODE_printIDS==True:\n",
    "            print(\"\\nThe following rhodopsins have this neighboring annot:\\n\")\n",
    "            print(\",\".join(III_list_annots_up_or_downstream_of_rhodopsin_starting_at_specific_pattern(DF2, down_pattern, 'down', 'rhodopsinIDs')))\n",
    "\n",
    "\n",
    "########### PART II. TABULATE PRESENCE/ABSENCE FOR MOST COMMON ANNOTS FLANKING RHODOPSIN ##################\n",
    "\n",
    "LIST_DoesPatternFlankRhodopsin = []  #For each pattern, contains  a DF of binary vals for presence or absence of that pattern v. each rhodopsin SeqID\n",
    "# Basically, contains 23 DFs, since there are 23 patterns.\n",
    "# And each DF has a row for each rhodopsin ID, and the cells can be valued 0 (absent) or 1 (present) \n",
    "\n",
    "print(\"\\n\\n~~~~~~~ Annotations Flanking Downstream ~~~~~~~~~\\n\")\n",
    "for down_pattern in list(DICT_downstream_annotPattern2Nickname.keys()):\n",
    "    _pattern_name = DICT_downstream_annotPattern2Nickname[down_pattern]\n",
    "    # List of rhodIDs to check for pattern\n",
    "    _LIST = DF2.loc[DF2['+1'].str.contains(down_pattern)]['rhodopsin_gene'].to_list()\n",
    "    _DF = pd.Series(data=np.nan, index=DF2['rhodopsin_gene'], name=_pattern_name).to_frame()\n",
    "    _DF[_pattern_name] = _DF.index.isin(_LIST).astype(int)\n",
    "    LIST_DoesPatternFlankRhodopsin.append(_DF)\n",
    "    print(_pattern_name, \"~\", down_pattern, \"~\", str(_DF[DICT_downstream_annotPattern2Nickname[down_pattern]].sum()))\n",
    "\n",
    "print(\"\\n\\n~~~~~~~~~ Annotations Flanking Upstream ~~~~~~~~~~~\\n\")\n",
    "for up_pattern in list(DICT_upstream_annotPattern2Nickname.keys()):\n",
    "    _pattern_name= DICT_upstream_annotPattern2Nickname[up_pattern]\n",
    "    # List of rhodIDs to check for pattern\n",
    "    _LIST = DF2.loc[DF2['-1'].str.contains(up_pattern)]['rhodopsin_gene'].to_list()\n",
    "    _DF = pd.Series(data=np.nan, index=DF2['rhodopsin_gene'], name=_pattern_name).to_frame()\n",
    "    _DF[DICT_upstream_annotPattern2Nickname[up_pattern]] = _DF.index.isin(_LIST).astype(int)\n",
    "    LIST_DoesPatternFlankRhodopsin.append(_DF)\n",
    "    print(_pattern_name, \"~\", up_pattern, \"~\", str(_DF[_pattern_name].sum()))\n",
    "    \n",
    "#### Save info to a Presence/Absence table (each row is a rhodopsin ID from GORG-Dark. Each cell is whether that rhodopsin is flanked by that annotation or not)\n",
    "\n",
    "DF_presence_absence = pd.concat(LIST_DoesPatternFlankRhodopsin, axis=1)\n",
    "DF_presence_absence = DF_presence_absence[LIST_cols_most_to_least_common]\n",
    "DF_presence_absence.to_csv(PATH_out)\n",
    "print(\"\\nSaved results to Presence/Absence table.\\nUse the table to the plot presence/absence of flanking genes on iToL phylogeny\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
